### PR TITLE
SNOW-3396010: has_aws_identity detector

### PR DESCRIPTION
**Pull request by sfc-gh-rsavenok**
## Summary
- Add `has_aws_identity` platform detector that calls STS `GetCallerIdentity` and accepts only IAM user / STS assumed-role ARNs suitable for workload identity federation.
- Introduce a `CallerIdentityProvider` trait and a `DetectionConfig` so detectors can be injected with fakes in tests and never hit real AWS.
- Run all detectors concurrently with a 200 ms per-detector timeout so a slow or unreachable STS endpoint cannot delay connection setup.
- Change `detect_platforms()` to `detect_platforms(&DetectionConfig)` and update the global-state caller to pass `DetectionConfig::default()`.
- Add `aws-sdk-sts = "1.86.0"` and move `tokio` into `[dev-dependencies]` with `test-util` to support paused-time tests.

## Context
Builds on `rsavenok/platform-detection-via-env` (#889), which added the env-based detector framework. This PR is the next step in SNOW-3396010: extending platform detection beyond environment variables to active identity probes, starting with AWS. The `CallerIdentityProvider` abstraction is designed so additional cloud-identity probes (e.g. GCP, Azure) can be plugged in the same way without changing `detect_platforms`.

## Test plan
- `cargo test -p sf_core telemetry::platform_detection` and `cargo test -p sf_core telemetry::aws_identity` — all pass locally.
- New tests cover:
  - `detects_has_aws_identity_when_provider_returns_wif_arn`
  - `rejects_when_sts_returns_no_arn`
  - `drops_when_provider_exceeds_timeout` (uses `#[tokio::test(start_paused = true)]` + `tokio::time::advance` to assert the 200 ms per-detector timeout fires)
  - `accepts_iam_user_and_assumed_role` and `rejects_malformed_arns` for `is_valid_arn_for_wif`.
- Existing env-detector tests were updated to use a fake provider that returns `None`, so the test suite never reaches the network.

### CODE REVIEW

**Comment by sfc-gh-rsavenok** [comment]
* **#893** <a href="https://app.graphite.com/github/pr/snowflakedb/universal-driver/893?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#892** <a href="https://app.graphite.com/github/pr/snowflakedb/universal-driver/892?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#891** <a href="https://app.graphite.com/github/pr/snowflakedb/universal-driver/891?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/snowflakedb/universal-driver/891?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#889** <a href="https://app.graphite.com/github/pr/snowflakedb/universal-driver/889?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#888** <a href="https://app.graphite.com/github/pr/snowflakedb/universal-driver/888?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#887** <a href="https://app.graphite.com/github/pr/snowflakedb/universal-driver/887?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: sf_core/src/apis/database_driver_v1/global_state.rs [hunk 1]
```diff
@@ -32,7 +32,9 @@ impl DatabaseDriverV1 {
     }
 
     pub async fn platforms(&self) -> &Vec<String> {
-        self.platforms.get_or_init(detect_platforms).await
+        self.platforms
+            .get_or_init(|| async { detect_platforms(&DetectionConfig::default()).await })
```
Comment (sfc-gh-pbulawa) [inline sf_core/src/apis/database_driver_v1/global_state.rs:69]: As the `DetectionConfig::default` is hardcoded here, we won't be able to inject a fake provider, right?

The test helper does not clear `AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY, AWS_SESSION_TOKEN, AWS_PROFILE, AWS_CONFIG_FILE, AWS_SHARED_CREDENTIALS_FILE, AWS_CONTAINER_CREDENTIALS_RELATIVE_URI, or AWS_WEB_IDENTITY_TOKEN_FILE`. And even if it did, `aws_config::defaults().load()` also resolves credentials from `~/.aws/credentials` and IMDS, neither of which are controlled by `temp_env`.

**Review by sfc-gh-pbulawa — CHANGES_REQUESTED** [review]

File: sf_core/src/telemetry/platform_detection.rs [hunk 2]
```diff
@@ -1,25 +1,68 @@
...
+        }
+    }
+}
 
-pub async fn detect_platforms() -> Vec<String> {
+/// Run all detectors concurrently with a per-detector [`DETECTION_TIMEOUT`].
+///
+/// Returns the names of only those detectors that succeeded. Order matches
+/// the detector list below so the serialized array is stable across runs.
+pub async fn detect_platforms(config: &DetectionConfig) -> Vec<String> {
```
Comment (sfc-gh-jszczerbinski) [inline sf_core/src/telemetry/platform_detection.rs:28]: This should return vec of enum values:
- Also consider implementing detect method: `def detect(self, detectionConfig)`

**Review by sfc-gh-jszczerbinski — CHANGES_REQUESTED** [review]

File: sf_core/src/apis/database_driver_v1/global_state.rs (hunk 1 above)
Comment (sfc-gh-rsavenok) [reply sf_core/src/apis/database_driver_v1/global_state.rs:69]: > As the DetectionConfig::default is hardcoded here, we won't be able to inject a fake provider, right?

We don't really need to test on this level. Provider is injected into the unit tests of platform detection:
```
let mut cfg = test_detection_config();
cfg.caller_identity_provider = Arc::new(FakeCallerIdentityProvider::new(None));
....
detect_platforms(&cfg).await
```

> The test helper does not clear  ....

Why does it need to clear them? It doesn't set them

> And even if it did, aws_config::defaults().load() also resolves credentials from ~/.aws/credentials and IMDS, neither of which are controlled by temp_env.

This is a good point. We have 2 integration tests which could accidentally detect cloud vm, but instead of complicating code with full dependency injection, I think it would be easier to change assertions

Comment (sfc-gh-rsavenok) [reply sf_core/src/apis/database_driver_v1/global_state.rs:69]: Adjusted integration tests - https://github.com/snowflakedb/universal-driver/pull/891/changes/4f45898269dc06e3bcf8e7dff6d0cd9e4b974769

Note that we now don't have a coverage for `[]`, so `skip_serializing_if = "Option::is_none"` won't be caught by tests in
```
 #[serde(rename = "PLATFORM")]
pub platforms: Vec<String>,
```

But I think it's overkill to add complex dependency injection to driverv1 just to be able to test this minor edge case

File: sf_core/src/telemetry/platform_detection.rs (hunk 2 above)
Comment (sfc-gh-rsavenok) [reply sf_core/src/telemetry/platform_detection.rs:28]: > This should return vec of enum values

What benefit do they give us? The value in code is used only once - in the detectors vector + tests. Seems like overkill to define an enum + string serialization for something that doesn't repeat outside of this file + 1 integration test

> Also consider implementing detect method:

Do you mean something like this? Seems like much more complex code than we have now ¯\_(ツ)_/¯
```
/// Identifiers for every platform the driver can report.
///
/// The `as_str` form is what gets serialized to the wire; keep it in sync
/// with the server-side expectations.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum Platform {
    IsAwsLambda,
    IsAzureFunction,
    IsGceCloudRunService,
    IsGceCloudRunJob,
    IsGithubAction,
    HasAwsIdentity,
}
impl Platform {
    pub fn as_str(self) -> &'static str {
        match self {
            Platform::IsAwsLambda => "is_aws_lambda",
            Platform::IsAzureFunction => "is_azure_function",
            Platform::IsGceCloudRunService => "is_gce_cloud_run_service",
            Platform::IsGceCloudRunJob => "is_gce_cloud_run_job",
            Platform::IsGithubAction => "is_github_action",
            Platform::HasAwsIdentity => "has_aws_identity",
        }
    }
}
/// Outcome of [`detect_platforms`].
///
/// Distinguishes "detection ran and found nothing" from "detection was
/// disabled by the kill switch" — both were previously squashed into the
/// same `Vec<String>`.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum DetectionOutcome {
    Disabled,
    Detected(Vec<Platform>),
}
#[derive(Clone)]
pub struct DetectionConfig {
    pub(crate) caller_identity_provider: Arc<dyn CallerIdentityProvider>,
}
impl Default for DetectionConfig {
    fn default() -> Self {
        Self {
            caller_identity_provider: Arc::new(StsCallerIdentityProvider),
        }
    }
}
/// A single platform detector.
///
/// Every detector receives the same [`DetectionConfig`] so new injectable
/// collaborators can be added without changing the detector list.
#[async_trait]
pub trait Detector: Send + Sync {
    fn platform(&self) -> Platform;
    async fn detect(&self, config: &DetectionConfig) -> bool;
}
struct AwsLambdaDetector;
#[async_trait]
impl Detector for AwsLambdaDetector {
    fn platform(&self) -> Platform {
        Platform::IsAwsLambda
    }
    async fn detect(&self, _config: &DetectionConfig) -> bool {
        env_non_empty("LAMBDA_TASK_ROOT")
    }
}
struct AzureFunctionDetector;
#[async_trait]
impl Detector for AzureFunctionDetector {
    fn platform(&self) -> Platform {
        Platform::IsAzureFunction
    }
    async fn detect(&self, _config: &DetectionConfig) -> bool {
        env_non_empty("FUNCTIONS_WORKER_RUNTIME")
            && env_non_empty("FUNCTIONS_EXTENSION_VERSION")
            && env_non_empty("AzureWebJobsStorage")
    }
}
struct GceCloudRunServiceDetector;
#[async_trait]
impl Detector for GceCloudRunServiceDetector {
    fn platform(&self) -> Platform {
        Platform::IsGceCloudRunService
    }
    async fn detect(&self, _config: &DetectionConfig) -> bool {
        env_non_empty("K_SERVICE")
            && env_non_empty("K_REVISION")
            && env_non_empty("K_CONFIGURATION")
    }
}
......
```

**Review by sfc-gh-jszczerbinski — APPROVED** [review]

**Review by sfc-gh-pbulawa — APPROVED** [review]
